### PR TITLE
Expose x-pack telemetry for failure store

### PR DESCRIPTION
**Pull request by gmarouli**
In this PR we extend the `data_streams` section in the usage API to include failure store stats. The current structure is:
```
  "data_streams" : {
    "available" : true,
    "enabled" : true,
    "data_streams" : 0,
    "indices_count" : 0
  }
```

We propose:
```
  "data_streams" : {
    "available" : true,
    "enabled" : true,
    "data_streams" : 0,
    "indices_count" : 0,
    "failure_store": {
      "enabled_count": 4,
      "failure_indices_count": 10
    }
  }
```

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java [hunk 1]
```diff
@@ -49,6 +50,12 @@ protected void innerXContent(XContentBuilder builder, Params params) throws IOEx
         super.innerXContent(builder, params);
         builder.field("data_streams", streamStats.totalDataStreamCount);
         builder.field("indices_count", streamStats.indicesBehindDataStream);
+        if (DataStream.isFailureStoreFeatureFlagEnabled()) {
+            builder.startObject("failure_store");
+            builder.field("enabled", streamStats.failureStoreEnabledDataStreamCount);
```
Comment (nielsbauman) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java:55]: Small bike shed (a.k.a. dog house?): I find it ever so slightly confusing that `data_streams.enabled` is a boolean but `data_streams.failure_store.enabled` is a number. What about doing `enabled_count`? I don't feel too strongly about this, so it wouldn't be a blocker for me.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java [hunk 2]
```diff
@@ -77,26 +84,50 @@ public static class DataStreamStats implements Writeable {
...
-        public DataStreamStats(long totalDataStreamCount, long indicesBehindDataStream) {
+        private final long failureStoreEnabledDataStreamCount;
+        private final long failureStoreIndicesCount;
+
+        public DataStreamStats(
+            long totalDataStreamCount,
+            long indicesBehindDataStream,
+            long failureStoreEnabledDataStreamCount,
+            long failureStoreIndicesCount
+        ) {
```
Comment (nielsbauman) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java:95]: If we convert this to a record, we can remove this constructor, `hashcode` and `equals`. Again, not a blocker for me if you feel it's out of scope.

File: x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java [hunk 3]
```diff
@@ -42,19 +43,30 @@ public void testDSXpackInfo() {
...
+            assertThat(logsTemplate, notNullValue());
+            assertThat(logsTemplate.get("name"), equalTo("logs"));
+            assertThat(((Map<?, ?>) logsTemplate.get("index_template")).get("data_stream"), notNullValue());
+        });
         assertBusy(() -> {
             Map<?, ?> logsTemplate = (Map<?, ?>) ((List<?>) getLocation("/_index_template/logs").get("index_templates")).get(0);
             assertThat(logsTemplate, notNullValue());
             assertThat(logsTemplate.get("name"), equalTo("logs"));
             assertThat(((Map<?, ?>) logsTemplate.get("index_template")).get("data_stream"), notNullValue());
         });
```
Comment (nielsbauman) [inline x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java:68]: Am I missing something or did you perhaps duplicate this `assertBusy` accidentally?

**Review by nielsbauman — COMMENTED** [review]
I left two optional comments and one less optional :D, other than that it LGTM :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java (hunk 1 above)
Comment (gmarouli) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java:55]: +1

File: x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java (hunk 3 above)
Comment (gmarouli) [reply x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java:68]: You are right :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java (hunk 2 above)
Comment (gmarouli) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java]: done

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java [hunk 4]
```diff
@@ -42,19 +43,24 @@ public void testDSXpackInfo() {
...
     public void testDSXpackUsage() throws Exception {
         Map<?, ?> dataStreams = (Map<?, ?>) getLocation("/_xpack/usage").get("data_streams");
         assertNotNull(dataStreams);
         assertTrue((boolean) dataStreams.get("available"));
         assertTrue((boolean) dataStreams.get("enabled"));
         assertThat(dataStreams.get("data_streams"), anyOf(equalTo(null), equalTo(0)));
-
+        assertThat(dataStreams, hasKey("failure_store"));
+        Map<String, Integer> failureStoreStats = (Map<String, Integer>) dataStreams.get("failure_store");
+        assertThat(failureStoreStats.get("enabled"), anyOf(equalTo(null), equalTo(0)));
```
Comment (nielsbauman) [inline x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java:55]:
```suggestion
        assertThat(failureStoreStats.get("enabled_count"), anyOf(equalTo(null), equalTo(0)));
```
The CI didn't fail, so adding `hasKey` checks might be valuable here?

File: x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java [hunk 5]
```diff
@@ -65,21 +71,29 @@ public void testDSXpackUsage() throws Exception {
...
         dataStreams = (Map<?, ?>) getLocation("/_xpack/usage").get("data_streams");
         assertNotNull(dataStreams);
         assertTrue((boolean) dataStreams.get("available"));
         assertTrue((boolean) dataStreams.get("enabled"));
-        assertThat("got: " + dataStreams, dataStreams.get("data_streams"), equalTo(1));
-        assertThat("got: " + dataStreams, dataStreams.get("indices_count"), equalTo(2));
+        assertThat("got: " + dataStreams, dataStreams.get("data_streams"), equalTo(2));
+        assertThat("got: " + dataStreams, dataStreams.get("indices_count"), equalTo(3));
+        failureStoreStats = (Map<String, Integer>) dataStreams.get("failure_store");
+        assertThat(failureStoreStats.get("enabled_count"), anyOf(equalTo(null), equalTo(1)));
```
Comment (nielsbauman) [inline x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java:86]: This should be:
```suggestion
        assertThat(failureStoreStats.get("enabled_count"), equalTo(1));
```
right?

**Review by nielsbauman — CHANGES_REQUESTED** [review]

File: x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java (hunk 4 above)
Comment (gmarouli) [reply x-pack/plugin/core/src/javaRestTest/java/org/elasticsearch/xpack/core/DataStreamRestIT.java:55]: Where was my mind when I was writing this oops....

**Review by nielsbauman — APPROVED** [review]
LGTM!

**Pull request merged**